pkg/util/xorm/sequence.go

```diff
@@ -3,65 +3,147 @@ package xorm
 import (
 	"context"
 	"database/sql"
-	"errors"
 	"fmt"
+	"sync"
 )
 
+// batchState represents the state of a sequence batch
+type batchState struct {
+	mu               sync.Mutex
+	nextValue        int64
+	lastValueInBatch int64
+}
+
 type sequenceGenerator struct {
 	db             *sql.DB
 	sequencesTable string
+	batchSize      int64
+
+	mu          sync.Mutex
+	batchStates map[string]*batchState // Track sequence batches per key (table:column)
 }
 
 func newSequenceGenerator(db *sql.DB) *sequenceGenerator {
 	return &sequenceGenerator{
 		db:             db,
 		sequencesTable: "autoincrement_sequences",
+		batchSize:      100, // Default batch size
+		batchStates:    make(map[string]*batchState),
 	}
 }
 
 func (sg *sequenceGenerator) Next(ctx context.Context, table, column string) (int64, error) {
-	// Current implementation fetches new value for each Next call.
 	key := fmt.Sprintf("%s:%s", table, column)
 
+	// First get or create the state with a global lock (only for map access)
+	sg.mu.Lock()
+	state, exists := sg.batchStates[key]
+	if !exists {
+		state = &batchState{
+			nextValue:        0,
+			lastValueInBatch: -1,
+		}
+		sg.batchStates[key] = state
+	}
+	sg.mu.Unlock() // Release global lock as soon as possible
+
+	// Now lock only the specific sequence state
+	state.mu.Lock()
+	defer state.mu.Unlock()
+
+	// If we've used all values in the current batch, get a new batch
+	if state.nextValue > state.lastValueInBatch {
+		start, end, err := sg.allocateNewBatch(ctx, key)
+		if err != nil {
+			return 0, err
+		}
+		state.nextValue = start
+		state.lastValueInBatch = end
+	}
+
+	// Return the next value from the batch
+	val := state.nextValue
+	state.nextValue++
+	return val, nil
+}
+
+// allocateNewBatch retrieves a new batch of sequence values from the database.
+// It returns the start and end values of the new batch on success.
+func (sg *sequenceGenerator) allocateNewBatch(ctx context.Context, key string) (start, end int64, err error) {
 	tx, err := sg.db.BeginTx(ctx, &sql.TxOptions{Isolation: sql.LevelSerializable})
 	if err != nil {
-		return 0, err
+		return 0, 0, err
 	}
 
-	// TODO: "FOR UPDATE". (Somehow this doesn't seem to be supported in Spanner emulator?)
-	r, err := tx.QueryContext(ctx, "SELECT next_value FROM "+sg.sequencesTable+" WHERE name = ?", key)
+	defer func() {
+		if err != nil {
```

Comment:
It's not clear to me from looking at the code, which `err` is used here :( Can we rename returned error to `retErr`? (Or distinguish them in some other way?)